feat: Add OTP 28+ compatibility with version-aware regex pattern handling

matthewsinclair opened the pull request:
Including:

- Implement OTP version detection in schema macro
- Add runtime schema compilation for OTP 28+ when regex patterns are present
- Maintain module attribute optimisation for non-regex schemas on all OTP versions
- Add deprecation warning for regex patterns on OTP 28+ with migration guidance
- Fix cast_parameters.ex to avoid regex in module attributes
- Update string test to check pattern source instead of regex struct equality

This ensures backward compatibility while providing a clear migration path for OTP 28+ users to move from regex patterns to string patterns.

matthewsinclair committed:
feat: Add OTP 28+ compatibility with version-aware regex pattern handling

- Implement OTP version detection in schema macro
- Add runtime schema compilation for OTP 28+ when regex patterns are present
- Maintain module attribute optimization for non-regex schemas on all OTP versions
- Add deprecation warning for regex patterns on OTP 28+ with migration guidance
- Fix cast_parameters.ex to avoid regex in module attributes
- Update string test to check pattern source instead of regex struct equality

This ensures backward compatibility while providing a clear migration path
for OTP 28+ users to move from regex patterns to string patterns.

mbuhot commented on lib/open_api_spex/schema.ex:
```diff
@@ -530,4 +530,24 @@ defmodule OpenApiSpex.Schema do
+    |> has_regex_pattern?()
+  end
+
+  def has_regex_pattern?(enumerable)
```
This guard is giving a warning CI:

```
==> open_api_spex
Compiling 80 files (.ex)
     warning: incompatible types:

         map() !~ [dynamic()]

     in expression:

         # lib/open_api_spex/schema.ex:544
         is_list(enumerable)

     where "enumerable" was given the type map() in:

         # lib/open_api_spex/schema.ex:544
         is_map_key(enumerable, :__struct__)

     where "enumerable" was given the type [dynamic()] in:

         # lib/open_api_spex/schema.ex:544
         is_list(enumerable)

     Conflict found at
     │
 544 │       when not is_struct(enumerable) and (is_list(enumerable) or is_map(enumerable)) do
     │                                           ~
     │
     └─ lib/open_api_spex/schema.ex:544:43: OpenApiSpex.Schema.has_regex_pattern?/1

Compilation failed due to warnings while using the --warnings-as-errors option
Error: Process completed with exit code 1.
```

zorbash commented on lib/open_api_spex.ex:
```diff
@@ -253,32 +253,59 @@ defmodule OpenApiSpex do
     prevent "... protocol has already been consolidated ..."
     compiler warnings.
   """
+  def should_use_runtime_compilation?(body) do
```
I don't think this should be a public function. It's also hijacking the `@doc` for `defmacro schema/2`.

matthewsinclair replied:
I'll submit a change to fix this, but that method is called from elsewhere, so I think it needs to be public, no? Unless I'm missing something. The @doc issue is easier to fix (I think). :)

matthewsinclair committed:
fix: Address PR feedback from mbuhot and zorbash

- Fix type incompatibility warning in has_regex_pattern?/1 by splitting function clauses
- Make should_use_runtime_compilation?/1 undocumented (@doc false)

Addresses feedback in https://github.com/open-api-spex/open_api_spex/pull/670#pullrequestreview-2917052825

zachdaniel commented:
This change isn't actually necessary anymore. OTP 28.1 will restore this functionality, and Elixir 1.19 will be delayed until that happens.

zachdaniel commented:
🤔 I don't. I just have it on good authority, but I could perhaps be wrong actually.

erinversfeldcodes commented:
Hi 👋  I'm new to the contributing side of open_api_spex. I came across issue [#688](https://github.com/open-api-spex/open_api_spex/issues/688) while looking for something to contribute to. It looks like it describes the same root problem this PR addresses, so it might be worth linking them.                                       

For my understanding, are we on track to merge this PR, or does the OTP 28.1 claim is still block things? One option for #688 in the meantime would be to document that ~r// patterns don't work inside the schema/1 macro with Elixir 1.18.4 + OTP 28, and point users toward the string form if upgrading isn't an option for them. That said, if this PR is on track to merge it seems more valuable to focus energy here rather than on a stopgap.
